### PR TITLE
Enable assembly compatibility on Solang parser

### PR DESCRIPTION
**Pull request by LucasSte**
This PR updates the Solang-parser to be compatible with inline assembly and does some refactoring.

### CODE REVIEW

File: solang-parser/src/pt.rs [hunk 1]
```diff
@@ -597,7 +597,8 @@ pub enum Statement {
     },
     Assembly {
         loc: Loc,
-        assembly: AssemblyBlock,
+        assembly: Vec<AssemblyStatement>,
+        dialect: Option<String>,
```
Comment (seanyoung) [inline solang-parser/src/pt.rs:601]: Very much a nitpick, but in the parse tree I would place fields in the order they appear in the original source, so `dialect` first. Also I think `statements` would be a better name than `assembly`.

File: solang-parser/src/pt.rs [hunk 2]
```diff
@@ -653,28 +658,52 @@ pub enum AssemblyStatement {
...
+    FunctionCall(Box<AssemblyFunctionCall>),
     Member(Loc, Box<AssemblyExpression>, Identifier),
     Subscript(Loc, Box<AssemblyExpression>, Box<AssemblyExpression>),
 }
 
-pub type AssemblyBlock = Vec<AssemblyStatement>;
+#[derive(Debug, PartialEq, Clone)]
+pub struct AssemblyTypedIdentifier {
+    pub loc: Loc,
+    pub identifier_name: Identifier,
```
Comment (seanyoung) [inline solang-parser/src/pt.rs:683]: When naming a field in a struct, repeating the name of the struct in fields in unnecessary. These fields can be called `name` and  `type`.

File: solang-parser/src/solidity.lalrpop [hunk 3]
```diff
@@ -752,8 +752,10 @@ NonIfStatement: Statement = {
     <l:@L> "unchecked" "{" <statements:Statement*> "}" <r:@R> => {
         Statement::Block { loc: Loc::File(file_no, l, r), unchecked: true, statements }
     },
-    <l:@L> "assembly" <assembly:AssemblyBlock> <r:@R> => {
-        Statement::Assembly { loc: Loc::File(file_no, l, r), assembly }
+    <l:@L> "assembly" <dialect:string?> <assembly:AssemblyBlock> <r:@R> => {
+        Statement::Assembly {
+         loc: Loc::File(file_no, l, r), assembly,
+          dialect: if let Some(dialect_name) = dialect {Some(dialect_name.to_owned())} else {None}}
```
Comment (seanyoung) [inline solang-parser/src/solidity.lalrpop:758]: You can write the if statement as `dialect.map(|dialect| dialect.to_owned())`, much shorter.

The indentation is askew.

File: solang-parser/src/solidity.lalrpop [hunk 4]
```diff
@@ -779,30 +781,21 @@ NonIfStatement: Statement = {
...
-    },
+     <l:@L> "{" <statement:AssemblyStatement> "}" <r:@R> => { AssemblyStatement::Block(Loc::File(file_no, l, r),
+ Box::new(statement))},
+ <AssemblyVariableDeclaration> => <>,
+ <AssemblyAssignment> => <>,
+ <AssemblyFunctionCall> => AssemblyStatement::FunctionCall(Box::new(<>)),
+ <AssemblyIf> => <>,
+ <AssemblyFor> => <>,
+ <AssemblySwitch> => <>,
+ <AssemblyFunctionDefinition> => <>,
```
Comment (seanyoung) [inline solang-parser/src/solidity.lalrpop:798]: Indentation

File: solang-parser/src/solidity.lalrpop [hunk 5]
```diff
@@ -866,23 +827,107 @@ AssemblyExpression0: AssemblyExpression = {
...
+       };
+     AssemblyStatement::FunctionDefinition(Box::new(AssemblyFunctionDefinition{
+    loc: Loc::File(file_no, l, r),
+    name,
+    params,
+    returns,
+    body
+ }))
+
+ }
```
Comment (seanyoung) [inline solang-parser/src/solidity.lalrpop:873]: indentation

File: solang-parser/src/solidity.lalrpop [hunk 6]
```diff
@@ -866,23 +827,107 @@ AssemblyExpression0: AssemblyExpression = {
...
+AssemblyTypedIdentifierList: Vec<AssemblyTypedIdentifier> = {
+ CommaOne<<AssemblyType>> => <>
+}
+
+AssemblyType: AssemblyTypedIdentifier = {
+    <l:@L> <var_name:AssemblyIdentifier> ":" <var_type:AssemblyIdentifier> <r:@R> =>
+ AssemblyTypedIdentifier{loc: Loc::File(file_no, l, r),
+  identifier_name: var_name, identifier_type: Some(var_type)},
+ <l:@L> <var_name:AssemblyIdentifier> <r:@R> =>
+ AssemblyTypedIdentifier{loc: Loc::File(file_no, l, r), identifier_name: var_name, identifier_type: None},
```
Comment (seanyoung) [inline solang-parser/src/solidity.lalrpop:886]: indentation

**Review by seanyoung — CHANGES_REQUESTED** [review]
Looks great just some minor issues.I hope they make sense.

File: solang-parser/src/pt.rs (hunk 2 above)
Comment (LucasSte) [reply solang-parser/src/pt.rs:683]: `type` is a reserved word in Rust. I can't name the field as `type`, so I am changing to `assembly_type`.

File: solang-parser/src/solidity.lalrpop [hunk 7]
```diff
@@ -752,8 +752,12 @@ NonIfStatement: Statement = {
     <l:@L> "unchecked" "{" <statements:Statement*> "}" <r:@R> => {
         Statement::Block { loc: Loc::File(file_no, l, r), unchecked: true, statements }
     },
-    <l:@L> "assembly" <assembly:AssemblyBlock> <r:@R> => {
-        Statement::Assembly { loc: Loc::File(file_no, l, r), assembly }
+    <l:@L> "assembly" <dialect:string?> <statements:AssemblyBlock> <r:@R> => {
```
Comment (seanyoung) [inline solang-parser/src/solidity.lalrpop:755]: Here you us the string token, which will give the string, but not its location. If the contents is something we do not expect, then in sema we would like to give a warning with the correct location. Therefore it is better to `StringLiteral`.

File: solang-parser/src/solidity.lalrpop [hunk 8]
```diff
@@ -812,48 +807,16 @@ AssemblyStatement: AssemblyStatement = {
...
-    },
-    AssemblyExpression0
-}
-
-AssemblyExpression0: AssemblyExpression = {
-    <a:@L> "true" <b:@R> => AssemblyExpression::BoolLiteral(Loc::File(file_no, a, b), true),
-    <a:@L> "false" <b:@R> => AssemblyExpression::BoolLiteral(Loc::File(file_no, a, b), false),
-    <l:@L> <n:number> <r:@R> => {
+AssemblyLiteral: AssemblyExpression = {
+<a:@L> "true" <t_type:(":" <AssemblyIdentifier>)?> <b:@R> => AssemblyExpression::BoolLiteral(Loc::File(file_no, a, b), true, t_type),
```
Comment (seanyoung) [inline solang-parser/src/solidity.lalrpop:817]: Indentation is a bit messy here.

File: solang-parser/src/solidity.lalrpop [hunk 9]
```diff
@@ -866,23 +829,110 @@ AssemblyExpression0: AssemblyExpression = {
...
+            body
+            }
+          )
+        )
+ }
+
+}
+
+AssemblyTypedIdentifierList: Vec<AssemblyTypedIdentifier> = {
+ CommaOne<<AssemblyType>> => <>
```
Comment (seanyoung) [inline solang-parser/src/solidity.lalrpop:881]: Indentation is a bit messy.

File: solang-parser/src/pt.rs (hunk 2 above)
Comment (seanyoung) [reply solang-parser/src/pt.rs]: This issue comes up in other parts of the parse tree. In other parts we have used `ty` instead.

**Pull request merged**